processing/src/main/java/io/druid/query/lookup/RegisteredLookupExtractionFn.java

```diff
@@ -49,93 +52,109 @@ public static RegisteredLookupExtractionFn create(
-    final LookupExtractorFactory factory = manager.get(lookup);
-    Preconditions.checkNotNull(factory, "lookup [%s] not found", lookup);
-    return new RegisteredLookupExtractionFn(
-        new LookupExtractionFn(
-            factory.get(),
-            retainMissingValue,
-            replaceMissingValueWith,
-            injective,
-            optimize
-        ),
-        lookup
-    );
+    this.manager = manager;
+    this.replaceMissingValueWith = replaceMissingValueWith;
+    this.retainMissingValue = retainMissingValue;
+    this.injective = injective;
+    this.optimize = optimize;
+    this.lookup = lookup;
   }
 
   @JsonProperty("lookup")
   public String getLookup()
   {
-    return name;
+    return lookup;
   }
 
   @JsonProperty("retainMissingValue")
   public boolean isRetainMissingValue()
   {
-    return delegate.isRetainMissingValue();
+    return retainMissingValue;
   }
 
   @JsonProperty("replaceMissingValueWith")
   public String getReplaceMissingValueWith()
   {
-    return delegate.getReplaceMissingValueWith();
+    return replaceMissingValueWith;
   }
 
   @JsonProperty("injective")
   public boolean isInjective()
   {
-    return delegate.isInjective();
+    return injective;
   }
 
   @JsonProperty("optimize")
   public boolean isOptimize()
   {
-    return delegate.isOptimize();
+    return optimize;
   }
 
   @Override
   public byte[] getCacheKey()
   {
-    return delegate.getCacheKey();
+    try (final ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
+      baos.write(StringUtils.toUtf8(getClass().getCanonicalName()));
+      baos.write(0xFF);
+      baos.write(StringUtils.toUtf8(getLookup()));
+      baos.write(0xFF);
+      baos.write(ensureDelegate().getCacheKey());
+      return baos.toByteArray();
+    }
+    catch (IOException e) {
+      throw Throwables.propagate(e);
+    }
   }
 
   @Override
   public String apply(Object value)
   {
-    return delegate.apply(value);
+    return ensureDelegate().apply(value);
   }
 
   @Override
   public String apply(String value)
   {
-    return delegate.apply(value);
+    return ensureDelegate().apply(value);
   }
 
   @Override
   public String apply(long value)
   {
-    return delegate.apply(value);
+    return ensureDelegate().apply(value);
   }
 
   @Override
   public boolean preservesOrdering()
   {
-    return delegate.preservesOrdering();
+    return ensureDelegate().preservesOrdering();
   }
 
   @Override
   public ExtractionType getExtractionType()
   {
-    return delegate.getExtractionType();
-  }
-
-  @Override
-  public String toString()
-  {
-    return "RegisteredLookupExtractionFn{" +
-           "delegate=" + delegate +
-           ", name='" + name + '\'' +
-           '}';
+    return ensureDelegate().getExtractionType();
+  }
+
+  private LookupExtractionFn ensureDelegate()
+  {
+    LookupExtractionFn delegate = delegateRef.get();
+    if (null == delegate) {
```

Comment:
Yes, this is a common synchronization pattern in the c world. Another option would be to create the object then do a CAS at the end against `null` if you feel that is cleaner.